gobblin-kafka/src/main/java/gobblin/kafka/schemareg/CachingKafkaSchemaRegistry.java

```diff
@@ -0,0 +1,123 @@
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.kafka.schemareg;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.Map;
+
+import com.google.common.base.Preconditions;
+
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * An implementation that wraps a passed in schema registry and caches interactions with it
+ * {@inheritDoc}
+ * */
+@Slf4j
+public class CachingKafkaSchemaRegistry<K,S> implements KafkaSchemaRegistry<K,S> {
+
+  private static final int DEFAULT_MAX_SCHEMA_REFERENCES = 10;
+  private final KafkaSchemaRegistry<K,S> _kafkaSchemaRegistry;
+  private final HashMap<String, Map<S, K>> _namedSchemaCache;
+  private final HashMap<K, S> _idBasedCache;
+  private final int _maxSchemaReferences;
+
+
+  public CachingKafkaSchemaRegistry(KafkaSchemaRegistry kafkaSchemaRegistry)
+  {
+    this(kafkaSchemaRegistry, DEFAULT_MAX_SCHEMA_REFERENCES);
+  }
+
+  /**
+   * Create a caching schema registry.
+   * @param kafkaSchemaRegistry: a schema registry that needs caching
+   * @param maxSchemaReferences: the maximum number of unique references that can exist for a given schema.
+   */
+  public CachingKafkaSchemaRegistry(KafkaSchemaRegistry kafkaSchemaRegistry, int maxSchemaReferences)
+  {
+    Preconditions.checkArgument(kafkaSchemaRegistry!=null, "KafkaSchemaRegistry cannot be null");
+    if (kafkaSchemaRegistry.hasInternalCache())
+    {
+      log.warn(new StringBuilder().append("The passed in KafkaSchemaRegistry ")
+          .append(kafkaSchemaRegistry.getClass().getCanonicalName())
+          .append(" already has a cache. Putting another cache on top isn't going to make things faster!").toString());
+    }
+    _kafkaSchemaRegistry = kafkaSchemaRegistry;
+    _namedSchemaCache = new HashMap<>();
+    _idBasedCache = new HashMap<>();
+    _maxSchemaReferences = maxSchemaReferences;
+  }
+
+  @Override
+  public K register(String name, S schema)
+      throws IOException, SchemaRegistryException {
+
+    Map<S, K> schemaIdMap;
+    if (_namedSchemaCache.containsKey(name))
+    {
+      schemaIdMap = _namedSchemaCache.get(name);
+    }
+    else {
+      // we really care about reference equality to de-dup using cache
+      // when it comes to registering schemas, so use an IdentityHashMap here
+      schemaIdMap = new IdentityHashMap<>();
+      _namedSchemaCache.put(name, schemaIdMap);
+    }
+
+    if (schemaIdMap.containsKey(schema))
+    {
+      return schemaIdMap.get(schema);
+    }
+    else
+    {
+      // check if schemaIdMap is getting too full
+      Preconditions.checkState(schemaIdMap.size() < _maxSchemaReferences, "Too many schema objects for " + name +". Cache is overfull.");
```

Comment:
This will throw IllegalStateException, whereas we should evict older entries or least frequently used. May be something on the lines of sized based cache here: https://github.com/google/guava/wiki/CachesExplained